runtime/sema/gen/main.go

```diff
@@ -107,9 +108,60 @@ func (g *generator) VisitFunctionDeclaration(decl *ast.FunctionDeclaration) (_ s
 		panic("static function declarations are not supported")
 	}
 
+	// Name
+
 	functionName := decl.Identifier.Identifier
 	fullTypeName := g.fullTypeName()
-	docString := g.declarationDocString(decl)
+
+	g.addDecls(
+		goConstDecl(
+			functionNameVarName(fullTypeName, functionName),
+			goStringLit(functionName),
+		),
+	)
+
+	// Type parameters
+
+	var typeParams map[string]string
+
+	if decl.TypeParameterList != nil {
+		typeParameters := decl.TypeParameterList.TypeParameters
+		typeParams = make(map[string]string, len(typeParameters))
+
+		for _, typeParameter := range typeParameters {
+			typeParameterName := typeParameter.Identifier.Identifier
+
+			var typeBound dst.Expr
+			if typeParameter.TypeBound != nil {
+				typeBound = typeExpr(
+					typeParameter.TypeBound.Type,
+					typeParams,
+				)
+			}
+
+			typeParams[typeParameterName] = functionTypeParameterVarName(
+				fullTypeName,
+				functionName,
+				typeParameterName,
+			)
+
+			g.addDecls(
+				goVarDecl(
+					functionTypeParameterVarName(
+						fullTypeName,
+						functionName,
+						typeParameterName,
+					),
+					typeParameterExpr(
+						typeParameterName,
+						typeBound,
+					),
+				),
+			)
+		}
+	}
```

Comment:
Would be good to extract this potion to a separate function, so this function doesn't get too long

Reply:
Good point! Refactored this whole function into sub-functions in 8e47559b6478bd4384ab3363d77e1c711c4ad734